### DISCUSSION TITLE
InterpolatedStringHandler and Indexers

### DISCUSSION THREAD

**Discussion by nietras** [dotnet/csharplang#6039] [Q&A]
I am trying to define an API that supports using `InterpolatedStringHandler` on `set_Item` for an indexer. This works but similar to how you can forward the instance of the type which a method belongs to by using `InterpolatedStringHandlerArgument` I'd like to do that for the indexer too. That is define the `set_Item` like `set_Item2` as shown below, and then have C# support that. For now I don't know how to actually apply `InterpolatedStringHandlerArgument` to `value` to get it to call `TraceInterpolatedStringHandler(int literalLength, int formattedCount, Row row)`. How can I bend things to get this working?

PS: One option would be to use a ThreadLocal variable 😱

```csharp
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Text;

public class C
{
    public void C_Test()
    {
        M(new Row(), 42);
    }

    public void M(Row row, int value)
    {
        row["C"] = $"Test {value}";
        row.set_Item2("C", $"Test {value}");
    }
}

public class Row
{
    readonly Dictionary<string, string> _nameToValue = new();
    internal readonly StringBuilder _stringBuilder = new();

    public TraceInterpolatedStringHandler this[string index]
    {
        set { _nameToValue[index] = value.GetFormattedText(); }
    }

    [SpecialName]
    public void set_Item2(string index, [InterpolatedStringHandlerArgument("")] ref TraceInterpolatedStringHandler value)
    {
        _nameToValue[index] = value.GetFormattedText();
    }
}

[InterpolatedStringHandler]
public ref struct TraceInterpolatedStringHandler
{
    // Storage for the built-up string
    readonly StringBuilder builder;

    public TraceInterpolatedStringHandler(int literalLength, int formattedCount)
    {
        builder = new StringBuilder(literalLength);
        Trace.WriteLine($"\tliteral length: {literalLength}, formattedCount: {formattedCount}");
    }

    public TraceInterpolatedStringHandler(int literalLength, int formattedCount, Row row)
    {
        row._stringBuilder.Clear();
        builder = row._stringBuilder;
        Trace.WriteLine($"\tliteral length: {literalLength}, formattedCount: {formattedCount} REUSE");
    }

    public void AppendLiteral(string s)
    {
        Trace.WriteLine($"\tAppendLiteral called: {{{s}}}");

        builder.Append(s);
        Trace.WriteLine($"\tAppended the literal string");
    }

    public void AppendFormatted<T>(T t)
    {
        Trace.WriteLine($"\tAppendFormatted called: {{{t}}} is of type {typeof(T)}");

        builder.Append(t?.ToString());
        Trace.WriteLine($"\tAppended the formatted object");
    }

    internal string GetFormattedText() => builder.ToString();
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAIAYAEGCMALANwCwAUNnvgHQAiUEA5jAPYDOCUAxu2ZbgI0ASgFcYXMAFMaAYVZgADqikAnAMpqAbjyl8KVIQBUpADwT8DAZjwAmHLIoBvCjjd4bGQg4D6JzgAUAJSu7i7k7pE4ALIBMFIA7jjCrAnBcDiEtkH8kQC+FKFuGJ7esSlJqqkZsAg4WhAoolIhEWFFUVUJANoARLK9ALo4ALw4ACS9/nVODU1Seb25UW5dNOxSCD4AkghSYLYB/b0Zk9M4s43NizkdBeT31nbJqc4dqlIQACasMCgAnjgGNwuL8IKp/gAeAhYDIwgB8OB8MAg0iMrAAalcpKMcPE0rc2m5amoUSgcB9vr8ATh1AhVLAmAAhUSoL5qJEwllsjljfHBSxEjw4IyqCDcKTbCRqRSsFAQPZfOkMmBMAASEBgXxQHIQAAsoOxujCcLB2aZBh1wis3BsZkiUWjMdjumazMMxnNmjQAOKbABirFUkAQipM5gFOHu+UKQu66kUUm4jBQADlUVJLUKSnhvHadnsDgETW7TBlulK9qpZfLFcrGRqtTrVABBVRMUTSCRHXpBYYfABmIrFEsrMrlCqkSvpDc12o5XpaVo6kWRGfRWPmrq17txi99AaDIbDZgQAruFEe5Ar0urE7rM9VjfnqizOcHOE4qlEIOH4slt41pO04quqc7Nm8QoAPRQbSCBBswOIDkGOD6jiwCsigCAALSiIo1BYO8nw/H8gL1qq3IoOyqg4BhPKqIKkQ5qK/5jnetZTuRYFNmoAS1DgKBQFWjQADJSKq+o1BIODIcGCqKvI4gIK0kTWjadFUbyeKJHBoGUdRASCcJKBiRJeqEjaf4SjQADqDJ7CJsBSAEkwADoIEZaiNAJ4lMPqIAXJ5Yomb5+p5BksnHlOikSAFTiRfJ0WsEpNzLG4V5MTYLGjoB96cY+3Evnx0lBaJoV6lJdQJaGSVKRkFQUqkKntEKkRrD4XKYdRcg6uC56te4GnUbi7WdfRaVRNlMh2UJUiOfELm9O5pXkjqZlxStpl+Xq4UyUeiVfDFSAXNVCnJRIeTJAAogAquoV29hNGXuDmXg4C2iiJlqjnGcW+C4OwzVuGpKxTbZ9lzU5i3uR9X1fD9XnktwjQ6l8cVOE47B5Njj2xpZQ1qDQsPiV8ASAxNkRgzNDlQ25CDEzuXyoXqOIrQRuNCs9xSlO9n0k4Gck1V8kJGPCARGKhQMXCu7hUxD83OXTDNfALUVM8jKCo+jTgINjl2GjgrBDgg/yJhcJuJkb4tBKlePqV1hPKwECAAPw0OiXHBBZNpy7NCvQ/TfOM8zSH7ULhvAAAVkmCAczG5AdCSqhkgROB+ggqsHeGZ5BKMiIE6o7usJ7hL3EAA==)

If you run `C_Test` this outputs something like:
```
	literal length: 5, formattedCount: 1
	AppendLiteral called: {Test }
	Appended the literal string
	AppendFormatted called: {42} is of type System.Int32
	Appended the formatted object
	literal length: 5, formattedCount: 1 REUSE
	AppendLiteral called: {Test }
	Appended the literal string
	AppendFormatted called: {42} is of type System.Int32
	Appended the formatted object
```

**Comment by nietras** [comment]
@ltrzesniewski reminded me that you can of course apply the attribute to the setter parameter by:
```csharp
    public TraceInterpolatedStringHandler this[string index]
    {
        [param: InterpolatedStringHandlerArgument("")]
        set { _nameToValue[index] = value.GetFormattedText(); }
    }
```
However, the compiler does not use this and still uses `TraceInterpolatedStringHandler(int literalLength, int formattedCount)`. 

This might be more of a C# language issue so should perhaps raise that there instead.

**Reply by ltrzesniewski** [reply]
I think you should open a Roslyn issue:

- Either interpolated string handler support on indexers was never intended and is a side effect of the current implementation
- Or indexers should be treated exactly like a call to the `set_Item` method, in which case `[param: InterpolatedStringHandlerArgument("")]` should be supported

In either case, this needs a clarification, as right now we have partial support. I suppose this applies to property setters too.
